jdaviz/configs/specviz/plugins/viewers.py

```diff
@@ -76,16 +78,61 @@ def on_mouse_or_key_event(self, data):
                 self.label_mouseover.value = ""
                 return
 
-            fmt = 'x={:+10.5e} y={:+10.5e}'
-            self.label_mouseover.pixel = fmt.format(x, y)
+            # Snap to the closest data point, not the actual mouse location.
+            sp = None
+            closest_i = 0
+            closest_wave = 0
+            closest_flux = 0
+            closest_maxsize = 0
+            closest_label = ''
+            closest_distance = None
+            for lyr in self.state.layers:
+                if ((not isinstance(lyr.layer, BaseData)) or (lyr.layer.ndim not in (1, 3))
+                        or (not lyr.visible)):
+                    continue
+
+                try:
+                    # TODO: Is there a way to cache this?
+                    sp = lyr.layer.get_object(
+                        cls=Spectrum1D, statistic=getattr(self.state, 'function', None))
```

Comment:
can you just use the arrays in the Lines object instead?  If that isn't accessible from the layer, we could instead loop through `self.figure.marks` (but would then need to filter out marks that are not corresponding to data entries, etc).

Reply:
I can try using the marks if you tell me how to properly grab them. I get lost in all the spectrum viewer internals. 

Will using marks be compatible with unit conversion (when it finally works) and what is the risk of it breaking if we refactor the display stuff inside spectrum viewer?

Reply:
Also, would using marks work for Cubeviz and Mosviz?

Reply:
You should be able to pull the units from the figure as well, but will also need to account for `as_steps` similar to how is done for the cubeviz slider https://github.com/spacetelescope/jdaviz/blob/66adb92b48ea38d81a13251395cee94bc601e2ba/jdaviz/configs/cubeviz/helper.py#L103-L106

Either way, it's probably useful to cache in the plugin and update whenever the underlying data changes, but that might require a significant refactor of the plugin code.  Right now, cubeviz is very noticeably laggy, I suspect because of this line and the extra expense of re-collapsing and creating a `Spectrum1D` object at every mouse move event)


https://user-images.githubusercontent.com/877591/210076688-424f8439-c3d7-4d32-8fba-c3b83b6b33f6.mov

Reply:
In that case... I wonder if part of this can be refactored out so all the viz with spectrum viewer can reuse this same code. What do you think?

https://github.com/spacetelescope/jdaviz/blob/66adb92b48ea38d81a13251395cee94bc601e2ba/jdaviz/configs/cubeviz/helper.py#L89

Reply:
I think the logic needs to be quite different here (can't just assume the first marks entry in specviz, for example).  But if there is any part that can be refactored to be shared, I'm all for that.  But with the complication of `as_steps`, I'm beginning to think just caching might be the cleanest way to go.

Reply:
The cleanest way to catch is probably deep upstream here though I have a feeling they won't want it because of extra memory usage...

https://github.com/glue-viz/glue/blob/2293019762136c8cda1ce67e9c89d5513013d7e7/glue/core/data_collection.py#L414